File: 808-number-of-matching-subsequences/number-of-matching-subsequences.cpp
class Solution {
public:
    bool isSubsequence(string &s, const string &word)
    {
        int i = 0, j = 0;
        while(i < s.length() && j < word.length())
        {
            if(s[i] == word[j])
            j++;
            i++;
        }
        return j == word.length();
    }

    int numMatchingSubseq(string s, vector<string>& words) 
    {
        unordered_map<string, int> map;
        for(string word : words)
        map[word]++;

        int count = 0;
        for(auto& [word, wordFrequency] : map)
        {
            if(isSubsequence(s, word))
            count += wordFrequency;
        }
        return count;
    }
};

File: 808-number-of-matching-subsequences/number-of-matching-subsequences.java
class Solution {
    private boolean isSubsequence(String s, String word) 
    {
        int i = 0, j = 0;
        while (i < s.length() && j < word.length()) 
        {
            if(s.charAt(i) == word.charAt(j))
            j++;
            i++;
        }
        return j == word.length();
    }

    public int numMatchingSubseq(String s, String[] words) 
    {
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }

        int count = 0;
        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            String word = entry.getKey();
            int wordFrequency = entry.getValue();
            if (isSubsequence(s, word)) {
                count += wordFrequency;
            }
        }
        return count;
    }
}